File: novachat-endstone/novachat_endstone/extension/base_extension.py
"""
Base extension module.

This module provides a base class for NovaChat extensions with common functionality.

Requirements: 10.2 - THE Endstone 扩展 SHALL 使用 Python 编写并放置在 extensions 目录
"""

import logging
from typing import TYPE_CHECKING, Optional

from novachat_endstone.extension.extension import NovaChatExtension
from novachat_endstone.extension.extension_meta import ExtensionMeta
from novachat_endstone.extension.extension_exception import ExtensionException

if TYPE_CHECKING:
    from novachat_endstone.plugin import NovaChatPlugin


class BaseExtension(NovaChatExtension):
    """
    Base class for NovaChat extensions.
    Provides default implementations for common extension functionality.
    
    Extensions can extend this class instead of implementing NovaChatExtension directly.
    """
    
    def __init__(self):
        """Initialize the base extension."""
        self._meta: Optional[ExtensionMeta] = None
        self._plugin: Optional['NovaChatPlugin'] = None
        self._logger: Optional[logging.Logger] = None
    
    def get_meta(self) -> ExtensionMeta:
        """
        Gets the extension metadata.
        
        Returns:
            The extension metadata
            
        Raises:
            ExtensionException: If metadata is not set
        """
        if self._meta is None:
            raise ExtensionException("Extension metadata not set")
        return self._meta
    
    def set_meta(self, meta: ExtensionMeta) -> None:
        """
        Sets the extension metadata.
        
        Args:
            meta: The extension metadata
        """
        self._meta = meta
        self._logger = logging.getLogger(f"NovaChat.Extension.{meta.id}")
    
    def get_plugin(self) -> Optional['NovaChatPlugin']:
        """
        Gets the NovaChat plugin instance.
        
        Returns:
            The plugin instance, or None if not set
        """
        return self._plugin
    
    def set_plugin(self, plugin: 'NovaChatPlugin') -> None:
        """
        Sets the NovaChat plugin instance.
        
        Args:
            plugin: The plugin instance
        """
        self._plugin = plugin
    
    @property
    def logger(self) -> logging.Logger:
        """
        Gets the extension's logger.
        
        Returns:
            The logger instance
        """
        if self._logger is None:
            self._logger = logging.getLogger("NovaChat.Extension")
        return self._logger
    
    def info(self, message: str) -> None:
        """
        Logs an info message.
        
        Args:
            message: The message to log
        """
        self.logger.info(message)
    
    def warning(self, message: str) -> None:
        """
        Logs a warning message.
        
        Args:
            message: The message to log
        """
        self.logger.warning(message)
    
    def error(self, message: str) -> None:
        """
        Logs an error message.
        
        Args:
            message: The message to log
        """
        self.logger.error(message)
    
    def debug(self, message: str) -> None:
        """
        Logs a debug message.
        
        Args:
            message: The message to log
        """
        self.logger.debug(message)

File: NovaChat/Bedrock/pmmp/src/NovaChat/Extension/BaseExtension.php
<?php

declare(strict_types=1);

namespace NovaChat\Extension;

use NovaChat\NovaChatPlugin;

/**
 * Base class for NovaChat extensions.
 * Provides default implementations for common extension functionality.
 * 
 * Extensions can extend this class instead of implementing NovaChatExtension directly.
 * 
 * Requirements: 10.1 - THE PMMP 扩展 SHALL 使用 PHP 编写并放置在 extensions 目录
 */
abstract class BaseExtension implements NovaChatExtension {
    
    private ?ExtensionMeta $meta = null;
    private ?NovaChatPlugin $plugin = null;
    
    /**
     * {@inheritdoc}
     */
    public function getMeta(): ExtensionMeta {
        if ($this->meta === null) {
            throw new ExtensionException("Extension metadata not set");
        }
        return $this->meta;
    }
    
    /**
     * {@inheritdoc}
     */
    public function setMeta(ExtensionMeta $meta): void {
        $this->meta = $meta;
    }
    
    /**
     * {@inheritdoc}
     */
    public function getPlugin(): ?NovaChatPlugin {
        return $this->plugin;
    }
    
    /**
     * {@inheritdoc}
     */
    public function setPlugin(NovaChatPlugin $plugin): void {
        $this->plugin = $plugin;
    }
    
    /**
     * Gets the extension's logger.
     * Logs messages with the extension name prefix.
     * 
     * @return \Logger|null the logger, or null if plugin not set
     */
    protected function getLogger(): ?\Logger {
        return $this->plugin?->getLogger();
    }
    
    /**
     * Logs an info message.
     * 
     * @param string $message the message to log
     */
    protected function info(string $message): void {
        $prefix = $this->meta !== null ? "[{$this->meta->getName()}] " : "";
        $this->getLogger()?->info($prefix . $message);
    }
    
    /**
     * Logs a warning message.
     * 
     * @param string $message the message to log
     */
    protected function warning(string $message): void {
        $prefix = $this->meta !== null ? "[{$this->meta->getName()}] " : "";
        $this->getLogger()?->warning($prefix . $message);
    }
    
    /**
     * Logs an error message.
     * 
     * @param string $message the message to log
     */
    protected function error(string $message): void {
        $prefix = $this->meta !== null ? "[{$this->meta->getName()}] " : "";
        $this->getLogger()?->error($prefix . $message);
    }
    
    /**
     * Logs a debug message if debug mode is enabled.
     * 
     * @param string $message the message to log
     */
    protected function debug(string $message): void {
        if ($this->plugin !== null && $this->plugin->isDebugMode()) {
            $prefix = $this->meta !== null ? "[{$this->meta->getName()}] " : "";
            $this->getLogger()?->info("[Debug] " . $prefix . $message);
        }
    }
}
